source/dotnet/Library/AdaptiveCards/AdaptiveTypedElementConverter.cs

```diff
@@ -66,84 +55,115 @@ public static void RegisterTypedElement<T>(string typeName = null)
-            if (typeName == null)
-            {
-                // Get value of this objectType's "Type" JsonProperty(Required)
-                var typeJsonPropertyRequiredValue = objectType.GetRuntimeProperty("Type")
-                    .CustomAttributes.Where(a => a.AttributeType == typeof(JsonPropertyAttribute)).FirstOrDefault()?
-                    .NamedArguments.Where(a => a.TypedValue.ArgumentType == typeof(Required)).FirstOrDefault()
-                    .TypedValue.Value.ToString();
-
-                // If this objectType does not require "Type" attribute, use the objectType's XML "TypeName" attribute
-                if (typeJsonPropertyRequiredValue == "0")
-                {
-                    typeName = objectType
-                        .GetRuntimeFields().Where(x => x.Name == "TypeName").FirstOrDefault()?
-                        .GetValue("TypeName").ToString();
-                }
-                else
-                {
-                    throw new AdaptiveSerializationException("Required property 'type' not found on adaptive card element");
-                }
-            }
+            string typeName = GetElementTypeName(objectType, jObject);
 
             if (TypedElementTypes.Value.TryGetValue(typeName, out var type))
             {
-                if (jObject.Value<string>("id") == null)
+                string objectId = jObject.Value<string>("id");
+                if (objectId == null)
                 {
                     if (typeof(AdaptiveInput).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
                     {
                         throw new AdaptiveSerializationException($"Required property 'id' not found on '{typeName}'");
                     }
                 }
-                else
+
+                // add id of element to ParseContext
+                AdaptiveInternalID internalID = AdaptiveInternalID.Current();
+                if (type != typeof(AdaptiveCard))
                 {
-                    string objectId = jObject.Value<string>("id");
-                    if (Ids.Contains(objectId))
-                    {
-                        throw new AdaptiveSerializationException($"Duplicate 'id' found: '{objectId}'");
-                    }
-                    else
-                    {
-                        Ids.Add(objectId);
-                    }
+                    internalID = AdaptiveInternalID.Next();
+                    ParseContext.PushElement(objectId, internalID);
                 }
 
                 var result = (AdaptiveTypedElement)Activator.CreateInstance(type);
                 try
                 {
                     serializer.Populate(jObject.CreateReader(), result);
                 }
-                catch (JsonSerializationException)
+                catch (JsonSerializationException) { }
+
+                // remove id of element from ParseContext
+                if (type != typeof(AdaptiveCard))
+                    ParseContext.PopElement();
```

Comment:
Should have a starting/closing brace even for single statement blocks

Reply:
yep :)

---
In reply to: [278323682](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278323682) [](ancestors = 278323682)